src/deploy/functions/checkIam.ts

```diff
@@ -104,3 +107,82 @@ export async function checkHttpIam(
   }
   logger.debug("[functions] found setIamPolicy permission, proceeding with deploy");
 }
+
+function reduceEventsToServices(services: Array<Service>, endpoint: backend.Endpoint) {
+  const service = serviceForEndpoint(endpoint);
+  if (!services.find((s) => s.name === service.name)) {
+    services.push(service);
+  }
+  return services;
+}
+
+/**
+ * Checks and sets the roles for specific resource service agents
+ * @param projectId project identifier
+ * @param want backend that we want to deploy
+ * @param have backend that we have currently deployed
+ */
+export async function ensureServiceAgentRoles(
+  projectId: string,
+  want: backend.Backend,
+  have: backend.Backend
+): Promise<void> {
+  // find new services
+  const wantServices = backend.allEndpoints(want).reduce(reduceEventsToServices, []);
+  const haveServices = backend.allEndpoints(have).reduce(reduceEventsToServices, []);
+  const newServices = wantServices.filter(
+    (wantS) => !haveServices.find((haveS) => wantS.name === haveS.name)
+  );
+  if (newServices.length === 0) {
+    return;
+  }
+  // get the full project iam policy
+  let policy: iam.Policy;
+  try {
+    policy = await getIamPolicy(projectId);
+  } catch (err) {
+    logger.warn(
```

Comment:
This should probably be logLabeledBullet with the "functions" prefix. I'd also ditch the royal "We". I'm also wondering if we can make this warning more actionable. Maybe something like:

```
Could not verify the necessary IAM configuration for the following newly-integrated services: pubsub.googleapis.com, storage.googleapis.com. Deployment may fail.
```

This would spawn StackOverflow support Qs where people ask what they can do if storage.googleapis.com cannot be validated and deployments are failing. Those Qs would be answered with the exact steps an admin needs to do to unblock deployments.

Reply:
Note that we'd need to avoid listing "noop" here.

Reply:
Sounds good. Since I ended up changing how the reduce function works (no longer adds `noop`s or `pubsub`) we can assume that every new service will actually require a binding update.